tensorstore/kvstore/s3/object_metadata.cc

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorstore/kvstore/s3/object_metadata.h"
+
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include "absl/status/status.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_split.h"
+#include "absl/time/time.h"
+#include "tensorstore/internal/http/http_response.h"
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/str_cat.h"
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+namespace {
+
+/// ifrom hashlib import md5; md5("".encode("utf-8")).hexdigest()
```

Comment:
ifrom

Reply:
Also, comment the use of this. Does it have to be an actual ETag? ie. is it interpreted elsewhere?
Why does it need to be quoted?

Reply:
Firstly, I've removed the `ObjectMetadata` struct/class as it was overkill -- only the `etag` is ever used, the other fields were redundant (except possibly `x-amz-version-id` in future for versioned S3 buckets).

Secondly, I've consolidated this StorageGeneration creation as well as the xml tag search functionality into `s3_metadata.{h,cc}`.

Then, the etag is simply is used to construct a StorageGeneration from the response headers in `s3_metadata.cc:StorageGenerationFromHeaders`, Its documented as follows:

```
/// An etag is a hash of the S3 object contained in double-quotes.
/// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag
/// This may or may not be an MD5 digest of the data
/// https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html
```

In general, an etag of the form `W/"<etag_value>"` is possible in the case of weak validation, but I've ignored it as S3 seems to only perform strong validation. `validate.cc:IsValidStorageGeneration` tests for the double quotes.

Thirdly, the empty hash etag documented above has been changed to `""` and moved into `s3_key_value_store.cc` closer to it's use point in `AddGenerationHeader`. The empty etag is used as an input to an `if_equal` header in the case of a `NoValue` `StorageGeneration`. `AddGenerationHeader` should align with `AddGenerationParam` in the `gcs` kvstore now.